clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3646,23 +3630,167 @@ static void handleSYCLIntelSchedulerTargetFmaxMhzAttr(Sema &S, Decl *D,
+  // If any of the expression is value dependent, test is invalid.
+  const auto *MaxGExpr = dyn_cast<ConstantExpr>(E);
+  const auto *WorkGExpr = dyn_cast<ConstantExpr>(E1);
+  if (!MaxGExpr || !WorkGExpr)
+    return ArgCheck::Unknown;
+
+  // Check the argument values.
+  return (MaxGExpr->getResultAsAPSInt() == 0 &&
+          WorkGExpr->getResultAsAPSInt() == 1)
+             ? ArgCheck::EqualToOne
+             : ArgCheck::NotEqualToOne;
+}
+
+void Sema::AddSYCLIntelMaxGlobalWorkDimAttr(Decl *D,
+                                            const AttributeCommonInfo &CI,
+                                            Expr *E) {
+  if (!E->isValueDependent()) {
+    // Validate that we have an integer constant expression and then store the
+    // converted constant expression into the semantic attribute so that we
+    // don't have to evaluate it again later.
+    llvm::APSInt ArgVal;
+    ExprResult Res = VerifyIntegerConstantExpression(E, &ArgVal);
+    if (Res.isInvalid())
+      return;
+    E = Res.get();
 
-  Expr *E = A.getArgAsExpr(0);
+    // This attribute must be in the range [0, 3].
+    if (ArgVal < 0 || ArgVal > 3) {
+      Diag(E->getBeginLoc(), diag::err_attribute_argument_out_of_range)
+          << CI << 0 << 3 << E->getSourceRange();
+      return;
+    }
 
-  if (!checkWorkGroupSizeValues(S, D, A)) {
-    D->setInvalidDecl();
-    return;
+    // Check to see if there's a duplicate attribute with different values
+    // already applied to the declaration.
+    if (const auto *DeclAttr = D->getAttr<SYCLIntelMaxGlobalWorkDimAttr>()) {
+      // If the other attribute argument is instantiation dependent, we won't
+      // have converted it to a constant expression yet and thus we test
+      // whether this is a null pointer.
+      if (const auto *DeclExpr = dyn_cast<ConstantExpr>(DeclAttr->getValue())) {
+        if (ArgVal != DeclExpr->getResultAsAPSInt()) {
+          Diag(CI.getLoc(), diag::warn_duplicate_attribute) << CI;
+          Diag(DeclAttr->getLoc(), diag::note_previous_attribute);
+        }
+        // Drop the duplicate attribute.
+        return;
+      }
+    }
+
+    // If the declaration has a SYCLIntelMaxWorkGroupSizeAttr or
+    // ReqdWorkGroupSizeAttr, check to see if they hold equal values
+    // (1, 1, 1) in case the value of SYCLIntelMaxGlobalWorkDimAttr
+    // equals to 0.
+    if (const auto *DeclAttr = D->getAttr<ReqdWorkGroupSizeAttr>()) {
+      ArgCheck Results[] = {AreArgValuesToOne(E, DeclAttr->getXDim()),
+                            AreArgValuesToOne(E, DeclAttr->getYDim()),
+                            AreArgValuesToOne(E, DeclAttr->getZDim())};
+
+      // Diagnose if any of the argument values of ReqdWorkGroupSizeAttr
+      // are not equal to one when value of SYCLIntelMaxGlobalWorkDimAttr
+      // equals to 0.
+      if (ArgVal == 0 && llvm::any_of(Results, [](ArgCheck Ch) {
+            return Ch == ArgCheck::NotEqualToOne;
+          })) {
```

Comment:
It may make sense to then put these two checks (for `ReqdWorkGroupSizeAttr` and `SYCLIntelMaxWorkGroupSizeAttr`) into their own function and call it from both the add and the merge function, if possible.

Reply:
> It may make sense to then put these two checks (for `ReqdWorkGroupSizeAttr` and `SYCLIntelMaxWorkGroupSizeAttr`) into their own function and call it from both the add and the merge function, if possible.

Yes, I agree. with you, @AaronBallman. I have created a separate function to reduce duplicated codes. Since we are checking SYCLIntelMaxGlobalWorkDimAttr value differently (from add function through ArgVal and Merge function through MergeExpr), i was not able to do this throught new function checkWorkGroupSizeAttrExpr().